### ISSUE TITLE
Ideas to real-time data analysis

### ISSUE THREAD

**Issue by pcuci** [nteract/initiatives#6]
Hi nteract team - what were the ideas for collaborative editing of notebooks?

RE: https://github.com/nteract/initiatives/tree/master/real-time-collab

I'm thinking meteor/ddp/web-sockets - though unclear how people would find each other across the web or behind a NAT

Do you have more architectural details on how to make collaborative real-time editing notebooks work?

**Comment by rgbkrk** [comment]
Thanks for posting this issue! It's about time we start outlining the architectural considerations for this.

At a high level, I want to get into having the notebook (and other models) created server side -- users will send and receive updates to the session.

<img width="699" alt="screen shot 2017-03-16 at 10 51 33 am" src="https://cloud.githubusercontent.com/assets/836375/24010956/93cb606c-0a36-11e7-9340-43cfc9355c96.png">

The "Session" server connects to the kernel rather than the users.

In order to support real-time, we've spent a lot of time building up our base application to have a central event bus (redux in this case) and for the UI to be a pure reflection of the state of the document (components should not have state unless local).

**Comment by pcuci** [comment]
Maybe something like https://en.wikipedia.org/wiki/Operational_transformation for merge conflict resolution?

I believe this is the tech used behind google docs live-merging

I'm imagining other people concurently editing the nteract notebook, we're co-coding together, executing code against the same kernel - I should see their names move around the cells back and forth as we edit the same cell even.

**Comment by rgbkrk** [comment]
Yep, I'd love to see that. Are you interested in working on it? Would you like to get familiar with nteract/nteract and the components we're exporting in https://github.com/nteract/nteract/tree/master/packages?

**Comment by pcuci** [comment]
@rgbkrk between a baby (on the way, yey!), a job, a couple startup projects, and friends - I'm positive I can make some time for this, haha!

I have a specific usecase I'm actively researching, and I keep coming back to the Jupyter stack as a possible prototype. It's a really boring usecase: to take _live smart notes_ during meetings, in real time.

Some groups I'm part of use google docs for comptes-rendus (great colaborative editing), which is ok, but I need something 10x more flexible to structure the data for post-processing, and this approach would require live-coding, since humans are unpredictable. This experiment is an intermediary step to an AI social bot, but I first need to train it with new data that doesn't yet exist!

cc: @David-Ro @lgeiger

**Comment by pcuci** [comment]
Maybe investigate [DDP](https://guide.meteor.com/methods.html) to keep state equal between all players. MDG plans to [publish their DDP packages to npm directly](https://github.com/meteor/meteor/blob/devel/Roadmap.md#full-transition-to-npm), other packages available also: https://www.npmjs.com/search?q=ddp
